QgsMapLayer::ReadFlag::FlagReadExtentFromXml to extend vector layer read capabilities to other layer types

rldhont opened the pull request:
## Description

In `QgsVectorLayer`, it is possible through the flag trust layer metadata to use extent read in XML instead of extent provided by the provider.

The `QgsMapLayer::ReadFlag::FlagReadExtentFromXml` can be used in layer readXML to read Extent from XML for raster, point cloud and other layers.

This is a refactoring of https://github.com/qgis/QGIS/pull/38662

Funded by Ifremer

rldhont commented:
@m-kuhn you have done some reviews on #38662, can you review this PR ? Thanks

nyalldawson commented:
Can we delay this till after 3.18? Does it need to be in 3.18.0?

rldhont commented:
I'd like to be in 3.18

rldhont commented:
Hi devs, I have rebased and fix conflict. Is anyone available to review it ?

rldhont commented:
Hi devs, I have rebase this branch and waiting for review.

Is review necessary for merging ?

m-kuhn commented on src/core/qgsmaplayer.h:
```diff
@@ -1001,6 +1002,22 @@ class CORE_EXPORT QgsMapLayer : public QObject
+     * \see dataSourceChanged()
+     * \since QGIS 3.20
+     */
+    virtual void setDataSource( const QString &dataSource, const QString &baseName, const QString &provider, const QgsDataProvider::ProviderOptions &options, QgsDataProvider::ReadFlags flags );
```
Can we make this method non-virtual and let it call an internal (private virtual) helper method instead?
Also can we deprecate the existing one?
The current approach where each subclass has the possibility to override 2 different setDataSource methods and we can't emit a centralized `dataSourceChanged` signal from the superclass isn't really API friendly.

m-kuhn reviewed (commented):
I like the goal of this pull request, speeding up layer loading by trusting cached meta data.

On the API level I am not completely convinced, we are starting to head into a direction similar to `readStyle` and `readSymbology` which are two virtual methods that call each other and it's very tricky to keep track of their interaction.

What I would do instead: have one (or more) non-virtual public `setDataSource` methods and `virtual private setDataSourcePrivate()` method where providers do their implementation. The parent one can then do some safety checks, emit signals etc while the child ones can do the provider specific work.

nyalldawson commented:
Mesh layers don't do anything with setDataSource currently (which is a bug, it prevents fixing the paths of these layers. Ping @vcloarec ) -- is this going to become a worse issue after these changed?

vcloarec commented:
@nyalldawson 
Indeed, mesh layers do not have `setDataSource` implemented and could not have the path fixed... I will look how fix that.
But I don't think this PR will change anything for mesh layer, extent is directly read from the provider in the the `QgsMeshLayer::extent()` method without using the `mExtent` of `QgsMapLayer`. (by the way, as point could layer)

rldhont commented:
Hi @m-kuhn @nyalldawson I have opened a new PR https://github.com/qgis/QGIS/pull/42890 to test the specific code for unvirtualize setDataSource.

rldhont committed:
QgsMapLayer::ReadFlag::FlagReadExtentFromXml to extend vector layer read capabilities to other layer types

In QgsVectorLayer, it is possible through the flag trust layer metadata to use extent read in XML
instead of extent provided by the provider.

The QgsMapLayer::ReadFlag::FlagReadExtentFromXml can be used inlayer readXML to read Extent from XML for raster, point cloud and other layers.

Funded by Ifremer

rldhont commented:
This PR in now based on #42890

rldhont commented:
Now that #42890 has been merged, the PR is ready for review and merge.

m-kuhn commented:
From an API point of view this looks good. I think it would be good to evaluate if the debug strings are useful for prod - and if yes add a bit more context to them or otherwise remove them.
It would be good to have someone from "team server" look over that. @pblottiere , @elpaso any chance you could give it a look?

elpaso commented on src/core/raster/qgsrasterlayer.cpp:
```diff
@@ -2032,6 +2036,22 @@ bool QgsRasterLayer::readXml( const QDomNode &layer_node, QgsReadWriteContext &c
+    // read extent
+    if ( mReadFlags & QgsMapLayer::FlagReadExtentFromXml )
+    {
+      QDomNode extentNode = layer_node.namedItem( QStringLiteral( "extent" ) );
```
```suggestion
      const QDomNode extentNode = layer_node.namedItem( QStringLiteral( "extent" ) );
```

elpaso commented on src/core/vector/qgsvectorlayer.cpp:
```diff
@@ -189,6 +189,7 @@ QgsVectorLayer::QgsVectorLayer( const QString &vectorLayerPath,
       providerFlags |= QgsDataProvider::FlagLoadDefaultStyle;
     }
     setDataSourcePrivate( vectorLayerPath, baseName, providerKey, providerOptions, providerFlags );
+    emit dataSourceChanged();
```
shouldn't this signal be emitted in setDataSourcePrivate ? (and removed from the other places where it's emitted)

elpaso replied:
> The `emit` is done in the `setDataSource` method not in `setDataSourcePrivate` to do not have to duplicate the `emit` in every layer type implementation. 

Ok, then.

> Do you think I have to use `setDataSource` from `QgsMapLayer` instead of `setDataSourcePrivate` ?

Isn't that already the case? https://github.com/qgis/QGIS/blob/master/src/core/qgsmaplayer.cpp#L1719

Sorry, I'm a little confused: why do you need to call it in the constructor in the first place?

rldhont replied:
In PR #42890, I have unvirtualize setDataSource, and I have replace setDataSoruce by setDataSourcePrivate. So unvirtualization has been firstly done in this PR and I probably forgot to apply this emit in #42890.

The setDataSource used in the QgsVectorLayer, before #42890, has `emit dataSourceChanged` before every `return`. I was thinking, it is important to keep `emit dataSourceChanged` here.

elpaso replied:
Any reason why you are calling `setDataSourcePrivate()` instead of `setDataSource()` here? It would take care of emitting the signal directly.

elpaso commented on src/core/raster/qgsrasterlayer.cpp:
```diff
@@ -674,13 +675,16 @@ void QgsRasterLayer::setDataProvider( QString const &provider, const QgsDataProv
-  // store the extent
-  setExtent( mbr );
+    // show the extent
+    QgsDebugMsgLevel( "Extent of layer: " + mbr.toString(), 4 );
```
```suggestion
```

elpaso reviewed (commented):
LGTM, just left a few minor notes

rldhont commented:
@elpaso I have fixed all the `const` and I have a question about `emit changes`.

rldhont committed:
Use setDataSource instead of setDataSourcePrivate

rldhont commented:
@elpaso I have fixed the setDataSourcePrivate use.

vcloarec commented:
For information, `setDataSourcePrivate` is now implemented also for mesh layer.

rldhont commented:
@nyalldawson @m-kuhn @elpaso all requested changes has been done. Is it good to you for merge ?

rldhont commented:
I have rebased and fixed conflict. This PR is ready for merge.

rldhont commented:
Hi devs, can this PR be merged in master for 3.20 ?

m-kuhn commented:
Yes, the sooner the better to have time for testing.
Thanks for this !
